src/bench_generic/input_generators.h

```diff
@@ -33,62 +33,139 @@
+using uint64 = std::uint64_t;
+
 using std_int64_t = std::int64_t;
 
 using fake_url_pair = std::pair<fake_url, fake_url>;
 
+using uint_std_pair32 = std::pair<std::uint32_t, std::uint32_t>;
+using uint_std_pair64 = std::pair<std::uint64_t, std::uint64_t>;
+
+using uint_tuple_pair32 = algo::uint_tuple<32, 32>;
+using uint_tuple_pair64 = algo::uint_tuple<64, 64>;
+
+struct use_pair {
+  template <size_t size>
+  using type = std::pair<algo::uint_t<size>, algo::uint_t<size>>;
+
+  template <typename T, typename U>
+  constexpr auto operator()(T x, U y) const {
+    return std::pair{x, y};
+  }
+};
+
+struct use_uint_tuple {
+  template <size_t size>
+  using type = algo::uint_tuple<size, size>;
+
+  template <typename T, typename U>
+  constexpr auto operator()(T x, U y) const {
+    algo::uint_tuple<algo::uint_bit_size<T>(), algo::uint_bit_size<U>()> res;
+    algo::set_at<0>(res, x);
+    algo::set_at<1>(res, y);
+    return res;
+  }
+};
+
 template <typename T>
-struct int_to_t {
-  std::enable_if_t<std::is_integral_v<T>, T> operator()(int x) const {
-    return x;
+struct generate_t {
+  template <typename Src>
+  std::enable_if_t<std::is_integral_v<T>, T> operator()(Src& src) const {
+    return static_cast<T>(src());
   }
 };
 
 template <>
-struct int_to_t<double> {
-  double operator()(int x) const { return 1.0 / x; }
+struct generate_t<double> {
+  template <typename Src>
+  double operator()(Src& src) const {
+    return 1.0 / src();
+  }
 };
 
 template <>
-struct int_to_t<fake_url> {
-  fake_url operator()(int x) const { return fake_url(x); }
+struct generate_t<fake_url> {
+  template <typename Src>
+  fake_url operator()(Src& src) const {
+    return fake_url(src());
+  }
 };
 
 template <>
-struct int_to_t<fake_url_pair> {
-  fake_url_pair operator()(int x) const {
+struct generate_t<fake_url_pair> {
+  template <typename Src>
+  fake_url_pair operator()(Src& src) const {
+    int x = src();
     return std::make_pair(fake_url(x), fake_url(x));
   }
 };
 
 template <>
-struct int_to_t<noinline_int> {
-  noinline_int operator()(int x) const { return noinline_int{x}; }
+struct generate_t<noinline_int> {
+  template <typename Src>
+  noinline_int operator()(Src& src) const {
+    return noinline_int{src()};
+  }
+};
+
+template <>
+struct generate_t<uint_std_pair32> {
+  template <typename Src>
+  uint_std_pair32 operator()(Src& src) const {
+    return uint_std_pair32{src(), src()};
+  }
+};
+
+template <>
+struct generate_t<uint_std_pair64> {
+  template <typename Src>
+  uint_std_pair64 operator()(Src& src) const {
+    return uint_std_pair64{src(), src()};
+  }
+};
+
+template <>
+struct generate_t<uint_tuple_pair32> {
+  template <typename Src>
+  uint_tuple_pair32 operator()(Src& src) const {
+    return uint_tuple_pair32{src(), src()};
+  }
+};
+
+template <>
+struct generate_t<uint_tuple_pair64> {
+  template <typename Src>
+  uint_tuple_pair64 operator()(Src& src) const {
+    return uint_tuple_pair64{src(), src()};
+  }
```

Comment:
All these specializations may be replaced with just one that uses `is_pair` trait or something similar.

Reply:
Ok.

Reply:
Done